lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp

```diff
@@ -64,16 +95,121 @@ simplifyAndReplace(SILInstruction *inst, CanonicalizeInstruction &pass) {
       [&pass](SILInstruction *deleted) { pass.killInstruction(deleted); });
 
   // Push the new instruction and any users onto the worklist.
-  pass.newUsers(result);
+  pass.notifyHasNewUsers(result);
   return nextII;
 }
+
+//===----------------------------------------------------------------------===//
+//                        Canonicalize Memory Operations
+//===----------------------------------------------------------------------===//
+
+// Given a load with multiple struct_extracts/tuple_extracts and no other uses,
+// canonicalize the load into several (struct_element_addr (load)) pairs.
+//
+// (struct_extract (load %base))
+//   ->
+// (load (struct_element_addr %base, #field)
+static SILBasicBlock::iterator
+splitAggregateLoad(LoadInst *loadInst, CanonicalizeInstruction &pass) {
+  // Keep track of the next iterator after any newly added or to-be-deleted
+  // instructions. This must be valid regardless of whether the pass immediately
+  // deletes the instructions or simply records them for later deletion.
+  auto nextII = std::next(loadInst->getIterator());
+
+  switch (loadInst->getOwnershipQualifier()) {
+  case LoadOwnershipQualifier::Unqualified:
+  case LoadOwnershipQualifier::Trivial:
+    break;
+  case LoadOwnershipQualifier::Take:
+  case LoadOwnershipQualifier::Copy:
+    // Skip ahead if the ownership qualifier isn't handled.
+    //
+    // FIXME: To handle Copy/Take, the code below needs to properly peek through
+    // and destroy borrow scopes.
+    return nextII;
+  }
+  struct ProjInstPair {
+    Projection proj;
+    SingleValueInstruction *svi;
+
+    // When sorting, just look at the projection and ignore the instruction.
+    bool operator<(const ProjInstPair &rhs) const { return proj < rhs.proj; }
+  };
+
+  // Add load projections to a projection list.
+  llvm::SmallVector<ProjInstPair, 8> projections;
+  for (auto *use : getNonDebugUses(loadInst)) {
+    auto *user = use->getUser();
+
+    // If we have any non SEI, TEI instruction, don't do anything here.
+    if (!isa<StructExtractInst>(user) && !isa<TupleExtractInst>(user))
+      return nextII;
+
+    auto extract = cast<SingleValueInstruction>(user);
+    projections.push_back({Projection(extract), extract});
+  }
+  // Avoid removing dead loads from raw SIL. That may weaken definite-init.
```

Comment:
I wonder if it could also cause problems for debug info. Not sure. May be worth expanding this comment.

Reply:
In RAW SIL, I simply don't remove dead loads, regardless of debug info. I'll move this check to the point at which we delete the load so it's independent of whether any projections were rewritten. i.e. the original load will never be deleted in RAW SIL, whether new projection loads were created or not. That way, this pass can't affect diagnostics:
```
_ = deadUseOfInitializedField // Still want an uninitialized use message
```